### PR TITLE
[TIMOB-24189] Support Visual Studio 2017

### PR DESCRIPTION
**Pull request by infosia**
[TIMOB-24189](https://jira.appcelerator.org/browse/TIMOB-24189)

~~_Requires https://github.com/appcelerator/windowslib/pull/65 and  https://github.com/appcelerator/windowslib/pull/66_~~
~~_Requires new windowslib_ - https://github.com/appcelerator/titanium_mobile/pull/8893~~

- Add support for Visual Studio 2017 installation

As of Visual Studio 2017, you can install multiple editions of Visual Studio (Community, Professional etc) and it requires new way to detect its instances. (See also  [this](https://github.com/Microsoft/vs-setup-samples) and [this](http://www.visualstudioextensibility.com/2016/11/23/some-implications-of-the-new-modular-setup-of-visual-studio-2017-for-vsx-developers/) ). 

So the problem for `windowslib` is that the result from `windowslib.visualstudio.detect` actually did not quite consider this situation and it uses MSBuild version number as a key to select Visual Studio such as `12.0` and `14.0`. You might think we could use `15.0` for VS 2017, but this does not work because we could have multiple installation of MSBuild that have same version number `15.0`. So for now we should have a way to choose one instance from multiple installation of Visual Studio 2017 _by its name_ such as `Visual Studio Community 2017 RC` and `Visual Studio Enterprise 2017 RC` etc.

So for now `visualstudio` output from `appc ti info -p windows -o json` is going to look like this:

```json
"visualstudio": {
        "Visual Studio Community 2017 RC": {
                "version": "Visual Studio Community 2017 RC",
                "supported": "true",
                "msbuildVersion": "15.0.26206.0",
                "path": "C:\\Program Files (x86)\\Microsoft Visual Studio\\2017\\Community",
                "wpsdk": null,
                "registryKey": null,
                "clrVersion": null,
                "selected": true,
                "vcvarsall": "C:\\PROGRA~2\\MIB055~1\\2017\\COMMUN~1\\VC\\AUXILI~1\\Build\\VCVARS~1.BAT"
        },
        "Visual Studio Professional 2017 RC": {
                "version": "Visual Studio Professional 2017 RC",
                "supported": "true",
                "msbuildVersion": "15.0.26206.0",
                "path": "C:\\Program Files (x86)\\Microsoft Visual Studio\\2017\\Professional",
                "wpsdk": null,
                "registryKey": null,
                "clrVersion": null,
                "selected": false,
                "vcvarsall": "C:\\PROGRA~2\\MIB055~1\\2017\\PRO~1\\VC\\AUXILI~1\\Build\\VCVARS~1.BAT"
        },
        "14.0": {
                "version": "14.0",
                "registryKey": "HKEY_CURRENT_USER\\Software\\Microsoft\\VisualStudio\\14.0_Config",
                "supported": true,
                "vcvarsall": "C:\\PROGRA~2\\MICROS~1.0\\VC\\VCVARS~1.BAT",
                "msbuildVersion": "14.0.25420.1",
                "wpsdk": null,
                "selected": false,
                "path": "C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\",
                "clrVersion": "v4.0.30319"
        },
        "12.0": {
                "version": "12.0",
                "registryKey": "HKEY_CURRENT_USER\\Software\\Microsoft\\VisualStudio\\12.0_Config",
                "supported": true,
                "vcvarsall": "C:\\PROGRA~2\\MICROS~1.0\\VC\\VCVARS~1.BAT",
                "msbuildVersion": "12.0.25420.1",
                "wpsdk": null,
                "selected": false,
                "path": "C:\\Program Files (x86)\\Microsoft Visual Studio 12.0\\",
                "clrVersion": "v4.0.30319"
        }
},
```

~~Note: Unfortunately, building against `Visual Studio 2017 RC` fails for now. According to the output from the command, the new `MSBuild` fails to parse specific configuration file `Microsoft.Net.CoreRuntime.targets` during compilation. I am guessing it's a bug of `MSBuild` and am trying find workaround for this.~~

```
> appc run -p windows --target ws-local --V  "Visual Studio Community 2017 RC" --wp-sdk 10.0 -l trace --build-only

[ERROR] C:\Program Files (x86)\MSBuild\15.0\.Net\CoreRuntime\Microsoft.Net.CoreRuntime.targets(193,11): error MSB4069: T
he "Microsoft.Build.Framework.ITaskItem[]" type of the "AssemblyList" parameter of the "ResourceHandlingTask" task is no
t supported by MSBuild. [C:\Users\xxx\.titanium\vsbuild\win10_classic\win10.x86\Win10Classic.vcxproj]
C:\Program Files (x86)\MSBuild\15.0\.Net\CoreRuntime\Microsoft.Net.CoreRuntime.targets(193,11): error MSB4026: The "Asse
mblyList=@(_FullAppLibrariesUnfiltered)" parameter for the "ResourceHandlingTask" task is invalid. [C:\Users\xxx/.titan
ium\vsbuild\win10_classic\win10.x86\Win10Classic.vcxproj]
[ERROR] C:\Program Files (x86)\MSBuild\15.0\.Net\CoreRuntime\Microsoft.Net.CoreRuntime.targets(192,9): error MSB4063: Th
e "ResourceHandlingTask" task could not be initialized with its input parameters.  [C:\Users\xxx\.titanium\vsbuild\win1
0_classic\win10.x86\Win10Classic.vcxproj]
```

### CODE REVIEW

**Comment by infosia** [comment]
So I'm still failing to make `msbuild` work. I suspect `msbuild` has issues because it fails even when you execute msbuild manually against UWP application project (that Visual Studio creates via "New Project -> Windows Universal C# app").

```
>C:\PROGRA~2\MICROS~1\2017\COMMUN~1\VC\AUXILI~1\Build\VCVARS~1.BAT x86 && MSBuild /v:d /p:Platform=x86 /p:Cofiguration=Release  "Projects\UWP1\UWP1.sln"
```

I'm still pushing this so that hopefully we could get something from community to this PR.

**Comment by garymathews** [comment]
@infosia Maybe we are missing dependencies from the installer?

**Comment by infosia** [comment]
So what's interesting is that when you open `.sln` file manually from Visual Studio 2017, you are able to build it. So maybe Visual Studio does something more than executing `msbuild` behind the scenes. maybe setting up configurations, or giving additional parameters to `msbuild`...?

**Comment by infosia** [comment]
Confirmed that this still fails with latest Visual Studio 2017 which is released officially today.

**Comment by infosia** [comment]
I was able to make it work locally by explicitly running Visual Studio IDE as a compiler (using [Devenv](https://msdn.microsoft.com/en-us/library/b20w810z.aspx)) behind the scenes. I guess Microsoft would not recommend doing it but we may need this workaround for now. Will update the PR soon. I *hope* use of Devenv also fixes the build issue we have been seeing in #958. Will update soon.

**Comment by infosia** [comment]
So `devenv` does compile VS project successfully, but it turns out it does not package the app like msbuild does during build 😞  (See also: [Visual Studio - Create App Package - Command Line](https://social.msdn.microsoft.com/Forums/en-US/4727e1e3-50f1-4835-8c09-131066963474/visual-studio-create-app-package-command-line?forum=toolsforwinapps)). I'm looking into what msbuild is doing for packaging.

**Comment by infosia** [comment]
I *think* I found another solution for this. I was able to make `msbuild` work without errors when I build it on the "Developer Command Prompt for VS 2017". I guess it's setting up dependencies to make msbuild work correctly. I think we can make sure to run it using `VsDevCmd.bat`. (See also [UWP store package build automation]( http://stackoverflow.com/questions/33794868/uwp-store-package-build-automation))

**Comment by infosia** [comment]
While working on this I found that some batch tools (including `vcvarsall.bat`) does not work well when it is executed by [8.3 filenames](https://en.wikipedia.org/wiki/8.3_filename). Pushing updates.

**Comment by infosia** [comment]
ok I am able to make this work on both mobile and Store app 👍 Pushing updates for windowslib too, this PR requires https://github.com/appcelerator/windowslib/pull/66

File: cli/commands/_build/compile.js [hunk 1]
```diff
@@ -69,7 +69,7 @@ function compileApp(next) {
 	}
 
 	// Use spawn directly so we can pipe output as we go
-	p = spawn(vsInfo.vcvarsall, [
+	p = spawn(path.join(vsInfo.path, 'Common7', 'Tools', 'VsDevCmd.bat'), [
```
Comment (garymathews) [inline cli/commands/_build/compile.js]: Maybe we should add this as a path in windowslib (`vsInfo.vsDevCmd`) so we only need to maintain it in one place?

**Comment by garymathews** [comment]
Awesome! 🎆 

Detects Visual Studio 2017
```
appc ti info -t windows
Appcelerator Command-Line Interface, version 6.1.0
Copyright (c) 2014-2017, Appcelerator, Inc.  All Rights Reserved.

Microsoft (R) Visual Studio
  14.0
    Path                      = C:\Program Files (x86)\Microsoft Visual Studio 14.0\
    CLR Version               = v4.0.30319
    MSBuild Version           = v14.0.25420.1
    Windows Phone SDKs        = not installed
  Visual Studio Community 2017 (selected)
    Path                      = C:\Program Files (x86)\Microsoft Visual Studio\2017\Community
    CLR Version               = null
    MSBuild Version           = v15.0.26228.9
    Windows Phone SDKs        = not installed

Microsoft (R) Windows Phone SDK
  10.0 (selected)
    Path                      = C:\Program Files (x86)\Windows Kits\10\
  8.1
    Path                      = C:\Program Files (x86)\Microsoft SDKs\Windows Phone\v8.1\
```

I can compile a Windows 10 store application fine, but when targeting ARM it fails:
**x86**
```
appc run -p windows -T ws-local -S 10.0 --build-only
Appcelerator Command-Line Interface, version 6.1.0
Copyright (c) 2014-2017, Appcelerator, Inc.  All Rights Reserved.

3/15/2017, 1:28:22 PM

Operating System
  Name                        = Microsoft Windows 10 Pro
  Version                     = 10.0.14393
  Architecture                = 64bit
  # CPUs                      = 16
  Memory                      = 34308431872

Node.js
  Node.js Version             = 7.6.0
  npm Version                 = 4.1.2

Titanium CLI
  CLI Version                 = 5.0.11

Titanium SDK
  SDK Version                 = 6.1.0.v20170309181051
  SDK Path                    = C:\ProgramData\Titanium\mobilesdk\win32\6.1.0.v20170309181051
  Target Platform             = windows

Command
  C:\Program Files\nodejs\node.exe C:\Users\gmathews\.appcelerator\install\6.1.0\package\node_modules\titanium\lib\titanium.js build run -p windows -T ws-local -S 10.0 --build-only --config-file C:\Users\gmathews\AppData\Local\Temp\build-1489584488054.json --log-level info --no-banner --project-dir C:\Users\gmathews\projects\testCase

[INFO]  Found Titanium module id=ti.cloud version=3.2.11 platform=commonjs deploy-type=development path=C:\ProgramData\Application Data\Titanium\modules\commonjs\ti.cloud\3.2.11
[INFO]  Building local Windows machine
[INFO]  Performing build only
[INFO]  Deploy type: development
[INFO]  Debugging disabled
[INFO]  Profiler disabled
[INFO]  Forcing rebuild: Titanium SDK path changed since last build
[INFO]    Was: C:\ProgramData\Titanium\mobilesdk\win32\6.1.0.v20170123100346\windows
[INFO]    Now: C:\ProgramData\Titanium\mobilesdk\win32\6.1.0.v20170309181051\windows
[INFO]  Using existing temporary pfx
[INFO]  JavaScript files need to be encrypted
[INFO]  Missing 7 app icons, generating missing icons
[INFO]    Square24x24Logo.png - Used for badge - size: 24x24
[INFO]    Square44x44Logo.png - Used for logo - size: 44x44
[INFO]    Square71x71Logo.png - Used for logo - size: 71x71
[INFO]    Square150x150Logo.png - Used for logo - size: 150x150
[INFO]    Logo.png - Used for logo - size: 150x150
[INFO]    StoreLogo.png - Used for logo - size: 50x50
[INFO]    SmallLogo.png - Used for logo - size: 30x30
[INFO]  Processing JavaScript files
[INFO]  Encrypting JavaScript files: C:\ProgramData\Titanium\mobilesdk\win32\6.1.0.v20170309181051\windows\titanium_prep.win32.exe "com.test" "C:\Users\gmathews\.titanium\vsbuild\testCase\Assets" "--file-listing" "C:\Users\gmathews\.titanium\vsbuild\testCase\titanium_prep_listing.txt"
[INFO]  Generating i18n files
[INFO]  Writing appxmanifest C:\Users\gmathews\.titanium\vsbuild\testCase\Package.store.appxmanifest.in
[INFO]  Writing appxmanifest C:\Users\gmathews\.titanium\vsbuild\testCase\Package.phone.appxmanifest.in
[INFO]  Writing appxmanifest C:\Users\gmathews\.titanium\vsbuild\testCase\Package.win10.appxmanifest.in
[INFO]  Generating Native Types
[INFO]  Setting up native modules for CMakeLists.txt...
[INFO]  Writing CMakeLists.txt: C:\Users\gmathews\.titanium\vsbuild\testCase\CMakeLists.txt
[INFO]  Running cmake at C:\ProgramData\Titanium\mobilesdk\win32\6.1.0.v20170309181051\windows\cli\vendor\cmake\bin\cmake.exe in directory C:\Users\gmathews\.titanium\vsbuild\testCase\win10.x86
[INFO]  -- The C compiler identification is MSVC 19.10.24930.0
[INFO]  -- The CXX compiler identification is MSVC 19.10.24930.0
[INFO]  -- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2017/Community/VC/Tools/MSVC/14.10.24930/bin/HostX86/x86/cl.exe
[INFO]  -- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2017/Community/VC/Tools/MSVC/14.10.24930/bin/HostX86/x86/cl.exe -- works
[INFO]  -- Detecting C compiler ABI info
[INFO]  -- Detecting C compiler ABI info - done
[INFO]  -- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2017/Community/VC/Tools/MSVC/14.10.24930/bin/HostX86/x86/cl.exe
[INFO]  -- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2017/Community/VC/Tools/MSVC/14.10.24930/bin/HostX86/x86/cl.exe -- works
[INFO]  -- Detecting CXX compiler ABI info
[INFO]  -- Detecting CXX compiler ABI info - done
[INFO]  -- Detecting CXX compile features
[INFO]  -- Detecting CXX compile features - done
[INFO]  -- Found GTest: C:/Users/gmathews/gtest-1.7.0-windows/msvc/gtest-md/Debug/gtest.lib
[INFO]  -- Looking for pthread.h
[INFO]  -- Looking for pthread.h - not found
[INFO]  -- Found Threads: TRUE
[INFO]  -- Found PythonInterp: C:/Python27/python.exe (found version "2.7.12")
[INFO]  -- Performing Test COMPILER_HAS_DEPRECATED_ATTR
[INFO]  -- Performing Test COMPILER_HAS_DEPRECATED_ATTR - Failed
[INFO]  -- Performing Test COMPILER_HAS_DEPRECATED
[INFO]  -- Performing Test COMPILER_HAS_DEPRECATED - Failed
[INFO]  -- Configuring done
[INFO]  -- Generating done
[INFO]  -- Build files have been written to: C:/Users/gmathews/.titanium/vsbuild/testCase/win10.x86
[INFO]  Targeting Windows SDK: 10.0
[INFO]  Running MSBuild on solution: C:\Users\gmathews\.titanium\vsbuild\testCase\win10.x86\TestCase.sln
[INFO]  Writing build manifest: C:\Users\gmathews\.titanium\vsbuild\testCase\build-manifest.json
[INFO]  Copying results back to project build directory
[INFO]  Project built successfully in 1m 51s 294ms
```

**ARM**
```
appc run -p windows -T wp-device -S 10.0 --build-only
Appcelerator Command-Line Interface, version 6.1.0
Copyright (c) 2014-2017, Appcelerator, Inc.  All Rights Reserved.

3/15/2017, 3:01:09 PM

Operating System
  Name                        = Microsoft Windows 10 Pro
  Version                     = 10.0.14393
  Architecture                = 64bit
  # CPUs                      = 16
  Memory                      = 34308431872

Node.js
  Node.js Version             = 7.6.0
  npm Version                 = 4.1.2

Titanium CLI
  CLI Version                 = 5.0.11

Titanium SDK
  SDK Version                 = 6.1.0.v20170309181051
  SDK Path                    = C:\ProgramData\Titanium\mobilesdk\win32\6.1.0.v20170309181051
  Target Platform             = windows

Command
  C:\Program Files\nodejs\node.exe C:\Users\gmathews\.appcelerator\install\6.1.0\package\node_modules\titanium\lib\titanium.js build run -p windows -T wp-device -S 10.0 --build-only --config-file C:\Users\gmathews\AppData\Local\Temp\build-1489590059728.json --log-level info --no-banner --project-dir C:\Users\gmathews\projects\testCase

[INFO]  Found Titanium module id=ti.cloud version=3.2.11 platform=commonjs deploy-type=test path=C:\ProgramData\Application Data\Titanium\modules\commonjs\ti.cloud\3.2.11
[INFO]  Building for Windows Phone device
[INFO]  Performing build only
[INFO]  Deploy type: test
[INFO]  Debugging disabled
[INFO]  Profiler disabled
[INFO]  Forcing rebuild: C:\Users\gmathews\.titanium\vsbuild\testCase\build-manifest.json does not exist
[INFO]  Using existing temporary pfx
[INFO]  JavaScript files need to be encrypted
[INFO]  Missing 7 app icons, generating missing icons
[INFO]    Square24x24Logo.png - Used for badge - size: 24x24
[INFO]    Square44x44Logo.png - Used for logo - size: 44x44
[INFO]    Square71x71Logo.png - Used for logo - size: 71x71
[INFO]    Square150x150Logo.png - Used for logo - size: 150x150
[INFO]    Logo.png - Used for logo - size: 150x150
[INFO]    StoreLogo.png - Used for logo - size: 50x50
[INFO]    SmallLogo.png - Used for logo - size: 30x30
[INFO]  Processing JavaScript files
[INFO]  Encrypting JavaScript files: C:\ProgramData\Titanium\mobilesdk\win32\6.1.0.v20170309181051\windows\titanium_prep.win32.exe "com.test" "C:\Users\gmathews\.titanium\vsbuild\testCase\Assets" "--file-listing" "C:\Users\gmathews\.titanium\vsbuild\testCase\titanium_prep_listing.txt"
[INFO]  Generating i18n files
[INFO]  Writing appxmanifest C:\Users\gmathews\.titanium\vsbuild\testCase\Package.store.appxmanifest.in
[INFO]  Writing appxmanifest C:\Users\gmathews\.titanium\vsbuild\testCase\Package.phone.appxmanifest.in
[INFO]  Writing appxmanifest C:\Users\gmathews\.titanium\vsbuild\testCase\Package.win10.appxmanifest.in
[INFO]  Generating Native Types
[INFO]  Setting up native modules for CMakeLists.txt...
[INFO]  Writing CMakeLists.txt: C:\Users\gmathews\.titanium\vsbuild\testCase\CMakeLists.txt
[INFO]  Running cmake at C:\ProgramData\Titanium\mobilesdk\win32\6.1.0.v20170309181051\windows\cli\vendor\cmake\bin\cmake.exe in directory C:\Users\gmathews\.titanium\vsbuild\testCase\win10.ARM
[WARN]  CMake Error at CMakeLists.txt:17 (project):
  A Windows Store component with CMake requires both the Windows Desktop SDK
  as well as the Windows Store '10.0' SDK.  Please make sure that you have
  both installed
[INFO]  -- Configuring incomplete, errors occurred!
See also "C:/Users/gmathews/.titanium/vsbuild/testCase/win10.ARM/CMakeFiles/CMakeOutput.log".
```

**Comment by infosia** [comment]
Updated to use `vsInfo.vsDevCmd` along with https://github.com/appcelerator/windowslib/pull/66

**Comment by garymathews** [comment]
CR: **PASS**
FT: **PASS**

###### TEST CASE
```
appc run -p windows -T wp-device
appc run -p windows -T ws-local
```

**Comment by infosia** [comment]
Note: In order to test this - you need to update windowslib.

1. Download [pre-built artifact from Jenkins](https://jenkins.appcelerator.org/job/appcelerator/job/titanium_mobile_windows/job/PR-959/8/display/redirect)
2. Then replace new [windowslib](https://github.com/appcelerator/titanium_mobile/pull/8893) with your YOUR_TITANIUM_SDK/node_modules/windowslib

**Comment by infosia** [comment]
This doesn't work on the lastest Visual Studio update 15.1 (26403.03).
Working on this again, removed "in-qe-testing" label.

**Comment by infosia** [comment]
So the issues we've been seeing is that

- Issue with long path [TIMOB-24569](https://jira.appcelerator.org/browse/TIMOB-24569)
- Unable to install app for StoreApp/Emulator/Device

**Comment by garymathews** [comment]
Could you cherry-pick this https://github.com/garymathews/titanium_mobile_windows/commit/58a732a52b251098fc8c6593f998cf016a0221c8 too?

**Comment by infosia** [comment]
Merged  https://github.com/garymathews/titanium_mobile_windows/commit/58a732a52b251098fc8c6593f998cf016a0221c8

**Comment by infosia** [comment]
I was able to make this work again by 

- Uninstall Visual Studio 2017, and then re-install it using [latest installer](https://www.visualstudio.com/thank-you-downloading-visual-studio/?sku=Community&rel=15).
- Then setting target platform version to `10.0.10586.0` other than `10.0.15063.0`

```xml
    <windows>
      <TargetPlatformVersion>10.0.10586.0</TargetPlatformVersion>
      <TargetPlatformMinVersion>10.0.10586.0</TargetPlatformMinVersion>
    </windows>
```

`10.0.14393.0` worked for me too.

```xml
    <windows>
      <TargetPlatformVersion>10.0.14393.0</TargetPlatformVersion>
      <TargetPlatformMinVersion>10.0.14393.0</TargetPlatformMinVersion>
    </windows>
```

**Comment by infosia** [comment]
@garymathews Could you test this PR again by using configuration described above? Thanks!

**Comment by garymathews** [comment]
@infosia That works, but would it be better to automatically detect the minimum SDK installed and use that?

**Comment by infosia** [comment]
> That works, but would it be better to automatically detect the minimum SDK installed and target that?

@garymathews Makes sense, pushed https://github.com/appcelerator/titanium_mobile_windows/pull/959/commits/8fc05186220fd431273ac01bec3f1ed682c78510 👍

**Comment by infosia** [comment]
I happened to have a issue with `vsInfo.vsDevCmd` when it is use with `spawn` on older version of Node.js. From what I can see Node.js may have incompatibility issue with spawn when it is used with space. (See also https://github.com/nodejs/node-v0.x-archive/issues/25895) 

Added "DO NOT MERGE" Label.

**Comment by garymathews** [comment]
Maybe we can get idea's from https://github.com/jdneo/node-phantom-simple/commit/8083e743cb43b4f0a49c4ae15a705e4241c9345e ?

**Pull request merged**